utils/zerovec/src/zerovec/mod.rs

```diff
@@ -684,37 +724,80 @@ where
     ///
     /// let bytes: &[u8] = &[0xD3, 0x00, 0x19, 0x01, 0xA5, 0x01, 0xCD, 0x01];
     /// let mut zerovec: ZeroVec<u16> = ZeroVec::parse_byte_slice(bytes).expect("infallible");
-    /// assert!(matches!(zerovec, ZeroVec::Borrowed(_)));
-    ///
-    /// zerovec.to_mut().push(12_u16.to_unaligned());
-    /// assert!(matches!(zerovec, ZeroVec::Owned(_)));
-    /// ```
-    pub fn to_mut(&mut self) -> &mut Vec<T::ULE> {
-        match self {
-            ZeroVec::Owned(ref mut vec) => vec,
-            ZeroVec::Borrowed(_) => {
-                let vec: Vec<T::ULE> = self.iter().map(T::to_unaligned).collect();
-                let new_self = ZeroVec::Owned(vec);
-                *self = new_self;
-                // recursion is limited since we are guaranteed to hit the Owned branch
-                self.to_mut()
-            }
-        }
+    /// assert!(!zerovec.is_owned());
+    ///
+    /// zerovec.with_mut(|v| v.push(12_u16.to_unaligned()));
+    /// assert!(zerovec.is_owned());
+    /// ```
+    pub fn with_mut<R>(&mut self, f: impl FnOnce(&mut Vec<T::ULE>) -> R) -> R {
+        // We're in danger if f() panics whilst we've moved a vector out of self;
+        // replace it with an empty dummy vector for now
+        let this = mem::take(self);
+        let mut vec = match this.into_cow() {
+            Cow::Owned(v) => v,
+            Cow::Borrowed(s) => s.into(),
+        };
+        let ret = f(&mut vec);
+        *self = Self::new_owned(vec);
+        ret
     }
 
+    /// Allows the ZeroVec to be mutated by converting it to an owned variant (if necessary)
+    /// and returning a slice to its backing buffer. [`Self::with_mut()`] allows for mutation
+    /// of the vector itself.
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
```

Comment:
Nit: Don't ignore this example. Make it so that it runs and passes.

Reply:
huh, i wodner why this was ignore initially